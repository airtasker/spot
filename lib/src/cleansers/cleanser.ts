import {
  ContractDefinition,
  defaultConfigDefinition
} from "../models/definitions";
import { ContractNode } from "../models/nodes";
import { cleanseApi } from "./nodes/api-cleanser";
import { cleanseConfig } from "./nodes/config-cleanser";
import { cleanseEndpoint } from "./nodes/endpoint-cleanser";

/**
 * Removes all additional parsing information associated with a parsed contract.
 *
 * @param contractNode a parsed contract
 */
export function cleanse(contractNode: ContractNode): ContractDefinition {
  const api = cleanseApi(contractNode.api.value);
  const config = contractNode.config
    ? cleanseConfig(contractNode.config.value)
    : defaultConfigDefinition;
  const endpoints = contractNode.endpoints.map(endpoint =>
    cleanseEndpoint(endpoint.value)
  );
  const types = contractNode.types;

  return { api, config, endpoints, types };
}
