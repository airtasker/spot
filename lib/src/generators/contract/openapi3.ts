import assertNever from "assert-never";
import YAML from "js-yaml";
import compact from "lodash/compact";
import pickBy from "lodash/pickBy";
import {
  ContractDefinition,
  DefaultResponseDefinition,
  EndpointDefinition,
  TypeDefinition
} from "../../models/definitions";
import { TypeKind } from "../../models/types";
import { resolveType } from "../../verifiers/utilities/type-resolver";
import { OpenAPI3SchemaType, openApi3TypeSchema } from "./openapi3-schema";

const SECURITY_HEADER_SCHEME_NAME = "securityHeader";

export function generateOpenApiV3(
  contractDefinition: ContractDefinition,
  format: "json" | "yaml"
) {
  const contract = openApiV3(contractDefinition);
  switch (format) {
    case "json":
      return JSON.stringify(contract, null, 2);
    case "yaml":
      return YAML.safeDump(contract, {
        skipInvalid: true // for undefined
      });
    default:
      throw assertNever(format);
  }
}

export function openApiV3(contractDefinition: ContractDefinition): OpenApiV3 {
  return {
    openapi: "3.0.0",
    info: {
      version: "0.0.0",
      title: contractDefinition.api.name,
      ...pickBy({ description: contractDefinition.api.description }),
      contact: {
        name: "TODO"
      }
    },
    ...(contractDefinition.api.securityHeader
      ? {
          security: [
            {
              [SECURITY_HEADER_SCHEME_NAME]: []
            }
          ]
        }
      : {}),
    paths: contractDefinition.endpoints.reduce<{
      [endpointPath: string]: {
        [method: string]: OpenAPIV3Operation;
      };
    }>((acc, endpoint) => {
      const openApiPath = endpoint.path.replace(/:(\w+)/g, "{$1}");
      acc[openApiPath] = acc[openApiPath] || {};
      acc[openApiPath][endpoint.method.toLowerCase()] = {
        operationId: endpoint.name,
        description: endpoint.description,
        tags: endpoint.tags,
        parameters: getParameters(contractDefinition.types, endpoint),
        ...(endpoint.request.body && {
          requestBody: {
            content: {
              "application/json": {
                schema: openApi3TypeSchema(
                  contractDefinition.types,
                  endpoint.request.body.type
                )
              }
            },
            description: endpoint.description || ""
          }
        }),
        responses: {
          ...(endpoint.defaultResponse
            ? {
                default: response(
                  contractDefinition.types,
                  endpoint.defaultResponse
                )
              }
            : {}),
          ...endpoint.responses.reduce<{
            [statusCode: string]: OpenAPIV3Body;
          }>((acc, responseNode) => {
            acc[responseNode.status.toString(10)] = response(
              contractDefinition.types,
              responseNode
            );
            return acc;
          }, {})
        }
      };
      return acc;
    }, {}),
    components: {
      schemas: contractDefinition.types.reduce<{
        [typeName: string]: OpenAPI3SchemaType;
      }>((acc, typeNode) => {
        acc[typeNode.name] = openApi3TypeSchema(
          contractDefinition.types,
          typeNode.type
        );
        return acc;
      }, {}),
      securitySchemes: contractDefinition.api.securityHeader
        ? {
            [SECURITY_HEADER_SCHEME_NAME]: {
              type: "apiKey",
              in: "header",
              name: contractDefinition.api.securityHeader.name,
              description: contractDefinition.api.securityHeader.description
            }
          }
        : {}
    }
  };
}

function getParameters(
  types: TypeDefinition[],
  endpoint: EndpointDefinition
): OpenAPIV3Parameter[] {
  const parameters = endpoint.request.pathParams
    .map(
      (pathParam): OpenAPIV3Parameter => {
        const resolvedType = resolveType(pathParam.type, types);
        const schemaType = openApi3TypeSchema(types, pathParam.type);
        return {
          in: "path",
          name: pathParam.name,
          description: pathParam.description,
          schema: schemaType,
          required: true,
          ...(resolvedType.kind === TypeKind.OBJECT
            ? {
                style: "deepObject",
                explode: true
              }
            : resolvedType.kind === TypeKind.ARRAY
            ? {
                style: "pipeDelimited",
                explode: true
              }
            : {})
        };
      }
    )
    .concat(
      endpoint.request.queryParams.map(
        (queryParam): OpenAPIV3Parameter => {
          const schemaType = openApi3TypeSchema(types, queryParam.type);
          return {
            in: "query",
            name: queryParam.name,
            description: queryParam.description,
            schema: schemaType,
            required: !queryParam.optional
          };
        }
      )
    )
    .concat(
      endpoint.request.headers.map(
        (header): OpenAPIV3Parameter => {
          const schemaType = openApi3TypeSchema(types, header.type);
          if ("type" in schemaType && schemaType.type === "object") {
            throw new Error(`Unsupported object type in header`);
          }
          return {
            in: "header",
            name: header.name,
            description: header.description,
            schema: schemaType,
            required: !header.optional
          };
        }
      )
    );
  return compact(parameters);
}

function response(
  types: TypeDefinition[],
  response: DefaultResponseDefinition
): OpenAPIV3Body {
  return {
    ...(response.body && {
      content: {
        "application/json": {
          schema: openApi3TypeSchema(types, response.body.type)
        }
      }
    }),
    headers: response.headers.reduce<{
      [header: string]: {
        description?: string;
        required: boolean;
        schema: OpenAPI3SchemaType;
      };
    }>((headerAcc, header) => {
      const schemaType = openApi3TypeSchema(types, header.type);
      headerAcc[header.name] = {
        description: header.description,
        required: !header.optional,
        schema: schemaType
      };
      return headerAcc;
    }, {}),
    description: response.description || ""
  };
}

export interface OpenApiV3 {
  openapi: "3.0.0";
  tags?: OpenAPIV3TagObject[];
  info: {
    version: string;
    title: string;
    description?: string;
    termsOfService?: string;
    contact?: {
      name?: string;
      url?: string;
      email?: string;
    };
    license?: {
      name: string;
      url?: string;
    };
  };
  servers?: Array<{
    url: string;
    description?: string;
  }>;
  security?: Array<{
    [securitySchemeName: string]: string[];
  }>;
  paths: {
    [endpointPath: string]: {
      [method: string]: OpenAPIV3Operation;
    };
  };
  components: {
    schemas: {
      [typeName: string]: OpenAPI3SchemaType;
    };
    securitySchemes?: {
      [securitySchemeName: string]: OpenAPIV3SecurityScheme;
    };
  };
}

export interface OpenAPIV3TagObject {
  name: string;
  description?: string;
}

export interface OpenAPIV3Operation {
  operationId: string;
  description?: string;
  tags?: string[];
  parameters: OpenAPIV3Parameter[];
  requestBody?: OpenAPIV3Body;
  responses: {
    default?: OpenAPIV3Body;
    // Note: we use | undefined because otherwise "default" would have to be required.
    [statusCode: string]: OpenAPIV3Body | undefined;
  };
}

export interface OpenAPIV3Parameter {
  in: "path" | "query" | "header";
  name: string;
  description?: string;
  required: boolean;
  schema: OpenAPI3SchemaType;
  explode?: boolean;
  style?:
    | "matrix"
    | "label"
    | "form"
    | "simple"
    | "spaceDelimited"
    | "pipeDelimited"
    | "deepObject";
}

export interface OpenAPIV3Body {
  content?: {
    "application/json": {
      schema: OpenAPI3SchemaType;
    };
  };
  headers?: {
    [header: string]: {
      description?: string;
      required: boolean;
      schema: OpenAPI3SchemaType;
    };
  };
  description: string;
}

// TODO: Consider adding support for other security schemes.
export type OpenAPIV3SecurityScheme = OpenApiV3SecurityScheme_ApiKey;

// tslint:disable-next-line:class-name
export interface OpenApiV3SecurityScheme_ApiKey {
  type: "apiKey";
  description?: string;
  name: string;
  in: "query" | "header" | "cookie";
}
