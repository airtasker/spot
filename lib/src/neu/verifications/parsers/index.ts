// @ts-ignore
import * as antlr4 from 'antlr4';
// @ts-ignore
import { QueryParamsLexer } from './QueryParamsLexer';
// @ts-ignore
import { QueryParamsParser } from './QueryParamsParser';

// let MyGrammarParser = require('./QueryParamsLexer').MyGrammarParser;
// let MyGrammarListener = require('./QueryParamsLexer').MyGrammarListener;

export function parseQueryParams(input:string) {
    let chars = new antlr4.InputStream(input);
    let lexer = new QueryParamsLexer(chars);
    let tokens  = new antlr4.CommonTokenStream(lexer);
    // console.log(tokens);
    let parser = new QueryParamsParser(tokens);
    parser.buildParseTrees = true;
    let tree = parser.querystring();
    console.log(tree.toStringTree(parser.ruleNames)); 
}
