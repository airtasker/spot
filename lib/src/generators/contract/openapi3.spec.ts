import { cleanse } from "../../cleansers/cleanser";
import { parse } from "../../parsers/parser";
import { generateOpenApiV3 } from "./openapi3";

const EXAMPLE_PATH = "./lib/src/__examples__/contract.ts";

describe("OpenAPI 3 generator", () => {
  test("produces valid code", async () => {
    const contractNode = await parse(EXAMPLE_PATH);
    const contractDefinition = cleanse(contractNode);
    expect(generateOpenApiV3(contractDefinition, "json")).toMatchSnapshot(
      "json"
    );
    expect(generateOpenApiV3(contractDefinition, "yaml")).toMatchSnapshot(
      "yaml"
    );
  });

  test("produces valid code when parameter serialization strategy is specified", async () => {
    const contractNode = await parse(
      "./lib/src/__examples__/contract-with-config.ts"
    );
    const contractDefinition = cleanse(contractNode);
    expect(generateOpenApiV3(contractDefinition, "json")).toMatchSnapshot(
      "json"
    );
    expect(generateOpenApiV3(contractDefinition, "yaml")).toMatchSnapshot(
      "yaml"
    );
  });
});
