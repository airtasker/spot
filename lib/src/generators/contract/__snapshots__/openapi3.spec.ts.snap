// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI 3 generator produces valid code when parameter serialization strategy is specified: json 1`] = `
"{
  \\"openapi\\": \\"3.0.0\\",
  \\"info\\": {
    \\"version\\": \\"0.0.0\\",
    \\"title\\": \\"company-api\\",
    \\"description\\": \\"This is the company API. It does cool things\\",
    \\"contact\\": {
      \\"name\\": \\"TODO\\"
    }
  },
  \\"security\\": [
    {
      \\"securityHeader\\": []
    }
  ],
  \\"paths\\": {
    \\"/_draft/company/{companyId}/users/{userId}\\": {
      \\"post\\": {
        \\"operationId\\": \\"GetUser\\",
        \\"description\\": \\"Retrieves a user in a company\\",
        \\"tags\\": [
          \\"Company\\",
          \\"User\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"companyId\\",
            \\"description\\": \\"company identifier\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          },
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"user identifier\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query\\",
            \\"description\\": \\"a demo query param\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": false
          },
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"x-auth-token\\",
            \\"description\\": \\"Auth Header\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"201\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/UserBody\\"
                }
              }
            },
            \\"headers\\": {
              \\"Location\\": {
                \\"description\\": \\"Location header\\",
                \\"required\\": true,
                \\"schema\\": {
                  \\"type\\": \\"string\\"
                }
              }
            },
            \\"description\\": \\"Successful creation of user\\"
          },
          \\"404\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/ErrorBody\\"
                }
              }
            },
            \\"headers\\": {},
            \\"description\\": \\"Bad request response\\"
          }
        }
      }
    },
    \\"/company/{companyId}/users\\": {
      \\"post\\": {
        \\"operationId\\": \\"CreateUser\\",
        \\"description\\": \\"Creates a user in a company\\",
        \\"tags\\": [
          \\"Company\\",
          \\"User\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"companyId\\",
            \\"description\\": \\"company identifier\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query\\",
            \\"description\\": \\"a demo query param\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": false
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query-array\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"string\\"
              }
            },
            \\"required\\": true,
            \\"explode\\": false,
            \\"style\\": \\"form\\"
          },
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"x-auth-token\\",
            \\"description\\": \\"Auth Header\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          }
        ],
        \\"requestBody\\": {
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"$ref\\": \\"#/components/schemas/CreateUserRequestBody\\"
              }
            }
          },
          \\"description\\": \\"Creates a user in a company\\"
        },
        \\"responses\\": {
          \\"201\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/UserBody\\"
                }
              }
            },
            \\"headers\\": {
              \\"Location\\": {
                \\"description\\": \\"Location header\\",
                \\"required\\": true,
                \\"schema\\": {
                  \\"type\\": \\"string\\"
                }
              }
            },
            \\"description\\": \\"Successful creation of user\\"
          },
          \\"400\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/ErrorBody\\"
                }
              }
            },
            \\"headers\\": {},
            \\"description\\": \\"Bad request response\\"
          },
          \\"default\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/ErrorBody\\"
                }
              }
            },
            \\"headers\\": {},
            \\"description\\": \\"\\"
          }
        }
      }
    }
  },
  \\"components\\": {
    \\"schemas\\": {
      \\"Address\\": {
        \\"type\\": \\"string\\"
      },
      \\"CreateUserRequestBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"data\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"firstName\\": {
                \\"type\\": \\"string\\"
              },
              \\"lastName\\": {
                \\"type\\": \\"string\\"
              },
              \\"age\\": {
                \\"type\\": \\"number\\",
                \\"format\\": \\"float\\"
              },
              \\"email\\": {
                \\"$ref\\": \\"#/components/schemas/Email\\"
              },
              \\"address\\": {
                \\"$ref\\": \\"#/components/schemas/Address\\"
              }
            },
            \\"required\\": [
              \\"firstName\\",
              \\"lastName\\",
              \\"age\\",
              \\"email\\",
              \\"address\\"
            ]
          }
        },
        \\"required\\": [
          \\"data\\"
        ]
      },
      \\"Email\\": {
        \\"type\\": \\"string\\"
      },
      \\"ErrorBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"name\\": {
            \\"type\\": \\"string\\"
          },
          \\"message\\": {
            \\"type\\": \\"array\\",
            \\"items\\": {
              \\"type\\": \\"string\\"
            }
          }
        },
        \\"required\\": [
          \\"name\\",
          \\"message\\"
        ]
      },
      \\"MessageOptions\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"newsletter\\": {
            \\"type\\": \\"boolean\\"
          }
        },
        \\"required\\": [
          \\"newsletter\\"
        ]
      },
      \\"Profile\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"private\\": {
            \\"type\\": \\"boolean\\"
          },
          \\"messageOptions\\": {
            \\"$ref\\": \\"#/components/schemas/MessageOptions\\"
          }
        },
        \\"required\\": [
          \\"private\\",
          \\"messageOptions\\"
        ]
      },
      \\"UserBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"data\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"firstName\\": {
                \\"type\\": \\"string\\"
              },
              \\"lastName\\": {
                \\"type\\": \\"string\\"
              },
              \\"profile\\": {
                \\"$ref\\": \\"#/components/schemas/Profile\\"
              }
            },
            \\"required\\": [
              \\"firstName\\",
              \\"lastName\\",
              \\"profile\\"
            ]
          }
        },
        \\"required\\": [
          \\"data\\"
        ]
      }
    },
    \\"securitySchemes\\": {
      \\"securityHeader\\": {
        \\"type\\": \\"apiKey\\",
        \\"in\\": \\"header\\",
        \\"name\\": \\"x-auth-token\\"
      }
    }
  }
}"
`;

exports[`OpenAPI 3 generator produces valid code when parameter serialization strategy is specified: yaml 1`] = `
"openapi: 3.0.0
info:
  version: 0.0.0
  title: company-api
  description: This is the company API. It does cool things
  contact:
    name: TODO
security:
  - securityHeader: []
paths:
  '/_draft/company/{companyId}/users/{userId}':
    post:
      operationId: GetUser
      description: Retrieves a user in a company
      tags:
        - Company
        - User
      parameters:
        - in: path
          name: companyId
          description: company identifier
          schema:
            type: string
          required: true
        - in: path
          name: userId
          description: user identifier
          schema:
            type: string
          required: true
        - in: query
          name: sample-query
          description: a demo query param
          schema:
            type: string
          required: false
        - in: header
          name: x-auth-token
          description: Auth Header
          schema:
            type: string
          required: true
      responses:
        '201':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserBody'
          headers:
            Location:
              description: Location header
              required: true
              schema:
                type: string
          description: Successful creation of user
        '404':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorBody'
          headers: {}
          description: Bad request response
  '/company/{companyId}/users':
    post:
      operationId: CreateUser
      description: Creates a user in a company
      tags:
        - Company
        - User
      parameters:
        - in: path
          name: companyId
          description: company identifier
          schema:
            type: string
          required: true
        - in: query
          name: sample-query
          description: a demo query param
          schema:
            type: string
          required: false
        - in: query
          name: sample-query-array
          schema:
            type: array
            items:
              type: string
          required: true
          explode: false
          style: form
        - in: header
          name: x-auth-token
          description: Auth Header
          schema:
            type: string
          required: true
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/CreateUserRequestBody'
        description: Creates a user in a company
      responses:
        '201':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserBody'
          headers:
            Location:
              description: Location header
              required: true
              schema:
                type: string
          description: Successful creation of user
        '400':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorBody'
          headers: {}
          description: Bad request response
        default:
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorBody'
          headers: {}
          description: ''
components:
  schemas:
    Address:
      type: string
    CreateUserRequestBody:
      type: object
      properties:
        data:
          type: object
          properties:
            firstName:
              type: string
            lastName:
              type: string
            age:
              type: number
              format: float
            email:
              $ref: '#/components/schemas/Email'
            address:
              $ref: '#/components/schemas/Address'
          required:
            - firstName
            - lastName
            - age
            - email
            - address
      required:
        - data
    Email:
      type: string
    ErrorBody:
      type: object
      properties:
        name:
          type: string
        message:
          type: array
          items:
            type: string
      required:
        - name
        - message
    MessageOptions:
      type: object
      properties:
        newsletter:
          type: boolean
      required:
        - newsletter
    Profile:
      type: object
      properties:
        private:
          type: boolean
        messageOptions:
          $ref: '#/components/schemas/MessageOptions'
      required:
        - private
        - messageOptions
    UserBody:
      type: object
      properties:
        data:
          type: object
          properties:
            firstName:
              type: string
            lastName:
              type: string
            profile:
              $ref: '#/components/schemas/Profile'
          required:
            - firstName
            - lastName
            - profile
      required:
        - data
  securitySchemes:
    securityHeader:
      type: apiKey
      in: header
      name: x-auth-token
"
`;

exports[`OpenAPI 3 generator produces valid code: json 1`] = `
"{
  \\"openapi\\": \\"3.0.0\\",
  \\"info\\": {
    \\"version\\": \\"0.0.0\\",
    \\"title\\": \\"company-api\\",
    \\"description\\": \\"This is the company API. It does cool things\\",
    \\"contact\\": {
      \\"name\\": \\"TODO\\"
    }
  },
  \\"security\\": [
    {
      \\"securityHeader\\": []
    }
  ],
  \\"paths\\": {
    \\"/_draft/company/{companyId}/users/{userId}\\": {
      \\"post\\": {
        \\"operationId\\": \\"GetUser\\",
        \\"description\\": \\"Retrieves a user in a company\\",
        \\"tags\\": [
          \\"Company\\",
          \\"User\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"companyId\\",
            \\"description\\": \\"company identifier\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          },
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"userId\\",
            \\"description\\": \\"user identifier\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query\\",
            \\"description\\": \\"a demo query param\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": false
          },
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"x-auth-token\\",
            \\"description\\": \\"Auth Header\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          }
        ],
        \\"responses\\": {
          \\"201\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/UserBody\\"
                }
              }
            },
            \\"headers\\": {
              \\"Location\\": {
                \\"description\\": \\"Location header\\",
                \\"required\\": true,
                \\"schema\\": {
                  \\"type\\": \\"string\\"
                }
              }
            },
            \\"description\\": \\"Successful creation of user\\"
          },
          \\"404\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/ErrorBody\\"
                }
              }
            },
            \\"headers\\": {},
            \\"description\\": \\"Bad request response\\"
          }
        }
      }
    },
    \\"/company/{companyId}/users\\": {
      \\"post\\": {
        \\"operationId\\": \\"CreateUser\\",
        \\"description\\": \\"Creates a user in a company\\",
        \\"tags\\": [
          \\"Company\\",
          \\"User\\"
        ],
        \\"parameters\\": [
          {
            \\"in\\": \\"path\\",
            \\"name\\": \\"companyId\\",
            \\"description\\": \\"company identifier\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          },
          {
            \\"in\\": \\"query\\",
            \\"name\\": \\"sample-query\\",
            \\"description\\": \\"a demo query param\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": false
          },
          {
            \\"in\\": \\"header\\",
            \\"name\\": \\"x-auth-token\\",
            \\"description\\": \\"Auth Header\\",
            \\"schema\\": {
              \\"type\\": \\"string\\"
            },
            \\"required\\": true
          }
        ],
        \\"requestBody\\": {
          \\"content\\": {
            \\"application/json\\": {
              \\"schema\\": {
                \\"$ref\\": \\"#/components/schemas/CreateUserRequestBody\\"
              }
            }
          },
          \\"description\\": \\"Creates a user in a company\\"
        },
        \\"responses\\": {
          \\"201\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/UserBody\\"
                }
              }
            },
            \\"headers\\": {
              \\"Location\\": {
                \\"description\\": \\"Location header\\",
                \\"required\\": true,
                \\"schema\\": {
                  \\"type\\": \\"string\\"
                }
              }
            },
            \\"description\\": \\"Successful creation of user\\"
          },
          \\"400\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/ErrorBody\\"
                }
              }
            },
            \\"headers\\": {},
            \\"description\\": \\"Bad request response\\"
          },
          \\"default\\": {
            \\"content\\": {
              \\"application/json\\": {
                \\"schema\\": {
                  \\"$ref\\": \\"#/components/schemas/ErrorBody\\"
                }
              }
            },
            \\"headers\\": {},
            \\"description\\": \\"\\"
          }
        }
      }
    }
  },
  \\"components\\": {
    \\"schemas\\": {
      \\"Address\\": {
        \\"type\\": \\"string\\"
      },
      \\"CreateUserRequestBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"data\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"firstName\\": {
                \\"type\\": \\"string\\"
              },
              \\"lastName\\": {
                \\"type\\": \\"string\\"
              },
              \\"age\\": {
                \\"type\\": \\"number\\",
                \\"format\\": \\"float\\"
              },
              \\"email\\": {
                \\"$ref\\": \\"#/components/schemas/Email\\"
              },
              \\"address\\": {
                \\"$ref\\": \\"#/components/schemas/Address\\"
              }
            },
            \\"required\\": [
              \\"firstName\\",
              \\"lastName\\",
              \\"age\\",
              \\"email\\",
              \\"address\\"
            ]
          }
        },
        \\"required\\": [
          \\"data\\"
        ]
      },
      \\"Email\\": {
        \\"type\\": \\"string\\"
      },
      \\"ErrorBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"name\\": {
            \\"type\\": \\"string\\"
          },
          \\"message\\": {
            \\"type\\": \\"array\\",
            \\"items\\": {
              \\"type\\": \\"string\\"
            }
          }
        },
        \\"required\\": [
          \\"name\\",
          \\"message\\"
        ]
      },
      \\"MessageOptions\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"newsletter\\": {
            \\"type\\": \\"boolean\\"
          }
        },
        \\"required\\": [
          \\"newsletter\\"
        ]
      },
      \\"Profile\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"private\\": {
            \\"type\\": \\"boolean\\"
          },
          \\"messageOptions\\": {
            \\"$ref\\": \\"#/components/schemas/MessageOptions\\"
          }
        },
        \\"required\\": [
          \\"private\\",
          \\"messageOptions\\"
        ]
      },
      \\"UserBody\\": {
        \\"type\\": \\"object\\",
        \\"properties\\": {
          \\"data\\": {
            \\"type\\": \\"object\\",
            \\"properties\\": {
              \\"firstName\\": {
                \\"type\\": \\"string\\"
              },
              \\"lastName\\": {
                \\"type\\": \\"string\\"
              },
              \\"profile\\": {
                \\"$ref\\": \\"#/components/schemas/Profile\\"
              }
            },
            \\"required\\": [
              \\"firstName\\",
              \\"lastName\\",
              \\"profile\\"
            ]
          }
        },
        \\"required\\": [
          \\"data\\"
        ]
      }
    },
    \\"securitySchemes\\": {
      \\"securityHeader\\": {
        \\"type\\": \\"apiKey\\",
        \\"in\\": \\"header\\",
        \\"name\\": \\"x-auth-token\\"
      }
    }
  }
}"
`;

exports[`OpenAPI 3 generator produces valid code: yaml 1`] = `
"openapi: 3.0.0
info:
  version: 0.0.0
  title: company-api
  description: This is the company API. It does cool things
  contact:
    name: TODO
security:
  - securityHeader: []
paths:
  '/_draft/company/{companyId}/users/{userId}':
    post:
      operationId: GetUser
      description: Retrieves a user in a company
      tags:
        - Company
        - User
      parameters:
        - in: path
          name: companyId
          description: company identifier
          schema:
            type: string
          required: true
        - in: path
          name: userId
          description: user identifier
          schema:
            type: string
          required: true
        - in: query
          name: sample-query
          description: a demo query param
          schema:
            type: string
          required: false
        - in: header
          name: x-auth-token
          description: Auth Header
          schema:
            type: string
          required: true
      responses:
        '201':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserBody'
          headers:
            Location:
              description: Location header
              required: true
              schema:
                type: string
          description: Successful creation of user
        '404':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorBody'
          headers: {}
          description: Bad request response
  '/company/{companyId}/users':
    post:
      operationId: CreateUser
      description: Creates a user in a company
      tags:
        - Company
        - User
      parameters:
        - in: path
          name: companyId
          description: company identifier
          schema:
            type: string
          required: true
        - in: query
          name: sample-query
          description: a demo query param
          schema:
            type: string
          required: false
        - in: header
          name: x-auth-token
          description: Auth Header
          schema:
            type: string
          required: true
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/CreateUserRequestBody'
        description: Creates a user in a company
      responses:
        '201':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UserBody'
          headers:
            Location:
              description: Location header
              required: true
              schema:
                type: string
          description: Successful creation of user
        '400':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorBody'
          headers: {}
          description: Bad request response
        default:
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorBody'
          headers: {}
          description: ''
components:
  schemas:
    Address:
      type: string
    CreateUserRequestBody:
      type: object
      properties:
        data:
          type: object
          properties:
            firstName:
              type: string
            lastName:
              type: string
            age:
              type: number
              format: float
            email:
              $ref: '#/components/schemas/Email'
            address:
              $ref: '#/components/schemas/Address'
          required:
            - firstName
            - lastName
            - age
            - email
            - address
      required:
        - data
    Email:
      type: string
    ErrorBody:
      type: object
      properties:
        name:
          type: string
        message:
          type: array
          items:
            type: string
      required:
        - name
        - message
    MessageOptions:
      type: object
      properties:
        newsletter:
          type: boolean
      required:
        - newsletter
    Profile:
      type: object
      properties:
        private:
          type: boolean
        messageOptions:
          $ref: '#/components/schemas/MessageOptions'
      required:
        - private
        - messageOptions
    UserBody:
      type: object
      properties:
        data:
          type: object
          properties:
            firstName:
              type: string
            lastName:
              type: string
            profile:
              $ref: '#/components/schemas/Profile'
          required:
            - firstName
            - lastName
            - profile
      required:
        - data
  securitySchemes:
    securityHeader:
      type: apiKey
      in: header
      name: x-auth-token
"
`;
